core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java

```diff
@@ -0,0 +1,151 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.index.percolator;
+
+import org.apache.lucene.document.Field;
+import org.apache.lucene.document.FieldType;
+import org.apache.lucene.index.*;
+import org.apache.lucene.queries.TermsQuery;
+import org.apache.lucene.search.*;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.index.mapper.ParseContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class QueryMetadataService {
+
+    public static String QUERY_METADATA = "_query_metadata_";
+    public static String QUERY_METADATA_FIELD_PREFIX = QUERY_METADATA + "field_";
+    public static String QUERY_METADATA_FIELD_UNKNOWN = QUERY_METADATA + "unknown";
+    public static FieldType QUERY_METADATA_FIELD_TYPE = new FieldType();
+
+    static {
+        QUERY_METADATA_FIELD_TYPE.setTokenized(false);
+        QUERY_METADATA_FIELD_TYPE.setIndexOptions(IndexOptions.DOCS);
+        QUERY_METADATA_FIELD_TYPE.freeze();
+    }
+
+    public void extractQueryMetadata(Query query, ParseContext.Document document) {
+        List<Term> queryTerms = new ArrayList<>();
+        extractQueryMetadata(query, queryTerms);
+        for (Term term : queryTerms) {
+            document.add(new Field(term.field(), term.bytes(), QUERY_METADATA_FIELD_TYPE));
+        }
+    }
+
+    public void extractQueryMetadata(Query query, List<Term> queryTerms) {
+        if (query instanceof TermQuery) {
+            Term term = ((TermQuery) query).getTerm();
+            queryTerms.add(new Term(QUERY_METADATA_FIELD_PREFIX + term.field(), term.bytes()));
+        } else if (query instanceof PhraseQuery) {
+            Term[] terms = ((PhraseQuery) query).getTerms();
+            if (terms.length == 0) {
+                return;
+            }
+
+            Term longestTerm = terms[0];
+            for (Term term : terms) {
+                if (longestTerm.bytes().length < term.bytes().length) {
+                    longestTerm = term;
+                }
+            }
+            queryTerms.add(new Term(QUERY_METADATA_FIELD_PREFIX + longestTerm.field(), longestTerm.bytes()));
+        } else if (query instanceof BooleanQuery) {
+            List<BooleanClause> clauses = ((BooleanQuery) query).clauses();
+            boolean noShouldClauses = true;
+            for (BooleanClause clause : clauses) {
+                if (clause.getOccur() == BooleanClause.Occur.SHOULD) {
+                    noShouldClauses = false;
+                    break;
+                }
+            }
```

Comment:
true, I completely misunderstood your comment.

On 8 October 2015 at 00:41, Adrien Grand <notifications@github.com> wrote:

> In
> core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java
> <https://github.com/elastic/elasticsearch/pull/13646#discussion_r41457315>
> :
>
> > +            Term longestTerm = terms[0];
> > +            for (Term term : terms) {
> > +                if (longestTerm.bytes().length < term.bytes().length) {
> > +                    longestTerm = term;
> > +                }
> > +            }
> > +            queryTerms.add(new Term(QUERY_METADATA_FIELD_PREFIX + longestTerm.field(), longestTerm.bytes()));
> > +        } else if (query instanceof BooleanQuery) {
> > +            List<BooleanClause> clauses = ((BooleanQuery) query).clauses();
> > +            boolean noShouldClauses = true;
> > +            for (BooleanClause clause : clauses) {
> > +                if (clause.getOccur() == BooleanClause.Occur.SHOULD) {
> > +                    noShouldClauses = false;
> > +                    break;
> > +                }
> > +            }
>
> For instance if you have SHOULD(a) MUST(b), the SHOULD clause is going to
> be completely optional, so we would just need to extract b. With the
> current code, we would extract both a and b. So instead of checking if
> there are optional clauses, I think we should check that there are no
> required clauses?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/13646/files#r41457315>.
>



-- 
Met vriendelijke groet,

Martijn van Groningen